### PR TITLE
Notifications for group invite are missed

### PR DESCRIPTION
**Pull request by aliaksandrb**
Currently we have no notifications for such use-cases:
- group member invites new users to a group
- network group invites the group to join it

Bug: 9676
Link: https://labs.riseup.net/code/issues/9676

### CODE REVIEW

File: test/functional/groups/invites_controller_test.rb [hunk 1]
```diff
@@ -65,4 +65,67 @@ def test_create_email_invite
...
+      post :create, group_id: @group.to_param,
+        recipients: "#{recipient.name}, #{@user.name}"
+    end
+
+    notice = RequestNotice.last
+    assert_equal recipient.id, notice.user_id
+    assert_equal 'request_to_join_us', notice.data[:title]
+  end
+
+  def test_invite_group_to_network_notifies_group_member
```
Comment (azul) [inline test/functional/groups/invites_controller_test.rb]: I'm not sure this really makes sense. If I'm the member of a network and i invite your group to the network I think you should be notified so you can react upon it.
This is quite cumbersome though because groups may be large and they don't necessarily have a council. So  we might have to notify all the members of the group that was invited. What do you think, @aliaksandrb ?

Comment (aliaksandrb) [inline test/functional/groups/invites_controller_test.rb]: Actually yes, if the group has no council, all members should be notified. Quite a lot of notifications for large group, but on the other hand we have not another possibility to invite a group rather than that.

How people from the closed group would know about the invite? (When you can not contact anybody personally to ask for council creation)

What about tests - not sure that I got the point, will double check it little bit later :)

**Comment by aliaksandrb** [comment]
Seems that I have got an idea of your notes, so updated the patch and tests.
Let me describe current implementation:

**Situation A:** group/network member invites other users to join it:
* notify all users from the recipient list who are not in this group already

**Situation B:** network invites group to join it:
* group has no council yet
  * notify all members who are not in this network already
* group has a council
  * notify all members from a council.

I have tried to cover this behavior with a tests, but maybe we need to cover more edge-cases if any.
Have though about mass-notification for group-members for **Situation B** above: actually do not think it could be a problem, because members of such groups also receive many notifications from constant page updates and actually it is not so often situation when network invites a group, so seems not a big deal for me :no_mouth:

As soon it would affect performance we could move all notification system into background jobs..

Ready for criticism :)

**Comment by azul** [comment]
@aliaksandrb, thanks a lot for the quick update. Great you thought about all these corner cases. I noticed one issue still though. I will comment on the line in question.

I agree that network membership requests are rare and we should not worry about the notification too much. Also agree that performance wise background jobs is the way to go.

File: app/controllers/groups/invites_controller.rb [hunk 2]
```diff
@@ -25,17 +25,24 @@ def create
...
       reqs << RequestToJoinUs.create(created_by: current_user,
         recipient: user, requestable: @group)
     end
+
     groups.each do |group|
-      reqs << RequestToJoinOurNetwork.create(created_by: current_user,
-        recipient: group, requestable: @group)
+      if council = group.council
+        reqs << RequestToJoinOurNetwork.create(created_by: current_user,
+          recipient: council, requestable: @group)
```
Comment (azul) [inline app/controllers/groups/invites_controller.rb]: The idea to only 'ask' the council is good. However if we use the council as the recipient of the request once the request is (approved)[https://github.com/riseuplabs/crabgrass-core/blob/master/app/models/requests/request_to_join_our_network.rb#L36] the council will be added to the network - not the group.

So I think what we want is to notify the people in the council but have the request still deal with the group.

Comment (aliaksandrb) [inline app/controllers/groups/invites_controller.rb]: > I confirmed that all group requests can only be approved by group admins. So only notifying them makes sense to me.

In this situation we send a notice to each group member, but only members of the council would be able to approve it, right?

That maybe it is better to update the `RequestNotice` model to handle it in such way:

```ruby
...
        elsif recipient.is_a? Group
          if recipient.council
            recipient.council.users.each do |user|
              create!(request: request, user: user)
            end
          else
            recipient.users.each do |user|
              create!(request: request, user: user)
            end
          end
...
```
where `request` - it is still `RequestToJoinOurNetwork` or any other, but notification would be created only to right users?
Because what is the reason to notify users that are not able to approve the request?

Comment (azul) [inline app/controllers/groups/invites_controller.rb]: sounds great! :star2:

**Comment by aliaksandrb** [comment]
I have updated the code. Now at `RequestNotice` model we check whether a group has a council and notify only its members if it is. Added unit test for such case as well.
